tests/Unit/TolerantVariableReferenceFinderTest.php

```diff
@@ -0,0 +1,198 @@
+        ];
+
+        yield 'scope: inside class method: inside anonumous function + use, click in use' => [
+            '<?php class C1 { function M1() { $var1 = 10; $f = function() use (<sr>$v<>ar1) { <sr>$var1 = 5; $var2 = <sr>$var1 + 10; } } }',
+        ];
+
+        yield 'scope: inside class method: inside anonumous function (no use), click inside' => [
+            '<?php class C1 { function M1() { $var1 = 10; $f = function($var1) { <sr>$v<>ar1 = 5; $var2 = <sr>$var1 + 10; } } }',
+        ];
+
+        yield 'scope: inside class method: inside anonumous function (no use), click outside' => [
+            '<?php class C1 { function M1() { $v<>ar1 = 10; $f = function($var1) { $var1 = 5; $var2 = $var1 + 10; } } }',
+        ];
+
+        yield 'scope: inside class method: inside anonumous class method' => [
+            '<?php '.
+                'class C1 { function M1() { '.
+                '$var = 1;'.
+                '$c = new class { function IM() { $v<>ar = 1; } } '.
+                ' } }',
+        ];
+
+        yield 'skip: static property access' => [
+            '<?php '.
+                'class C1 { static $prop1; function M1() { self::$pro<>p1 = 5; $var4 = self::$prop1; } }',
+        ];
+
+        yield 'skip: static property declaration' => [
+            '<?php '.
+                'class C1 { static $pr<>op1; function M1() { self::$prop1 = 5; $var4 = self::$prop1; } }',
+        ];
+
+        yield 'skip: instance property declaration' => [
+            '<?php '.
+                'class C1 { public $pr<>op1; function M1() { $this->prop1 = 5; $var4 = $this->prop1; } }',
+        ];
+    }
+
+    private static function offsetsFromSource(string $source, ?string $uri): array
+    {
+        $textDocumentUri = $uri !== null ? TextDocumentUri::fromString($uri) : null;
+        $results = preg_split("/(<>|<sr>|<mr>)/u", $source, null, PREG_SPLIT_DELIM_CAPTURE | PREG_SPLIT_NO_EMPTY);
+        
+        $referenceLocations = [];
+        $selectionOffset = null;
+
+        if (is_array($results)) {
+            $newSource = "";
+            $offset = 0;
+            foreach ($results as $result) {
+                if ($result == "<>") {
+                    $selectionOffset = $offset;
+                    continue;
+                }
+                
+                if ($result == "<sr>") {
+                    $referenceLocations[] = PotentialLocation::surely(
+                        new Location($textDocumentUri, ByteOffset::fromInt($offset))
+                    );
+                    continue;
+                }
+                if ($result == "<mr>") {
+                    $referenceLocations[] = PotentialLocation::maybe(
+                        new Location($textDocumentUri, ByteOffset::fromInt($offset))
+                    );
+                    continue;
+                }
+                
+                $newSource .= $result;
+                $offset += mb_strlen($result);
+            }
+        } else {
```

Comment:
can we invert this:

```
if (!is_array($blah)) {
    throw new Exception(...');
}

// do the stuff here
```

Reply:
Sure. I actually rather like this approach, but you see it faster than I do.